### PR TITLE
Merge Flex JAR feature branch into master

### LINKED ISSUE #2391: FlexDeployCommandHandler throws wrong exception
`IllegalStateException` is incorrect here. This should be a checked exception, possibly a `CoreException`, since it comes from a problem outside the code. 

```
  @Override
  protected StagingDelegate getStagingDelegate(IProject project) {
    String appYamlPath = new FlexDeployPreferences(project).getAppYamlPath();
    IFile appYaml = project.getFile(appYamlPath);
    if (!appYaml.exists()) {
      throw new IllegalStateException("Invalid app.yaml path from project preferences.");
    }
    IPath appEngineDirectory = appYaml.getParent().getLocation();
    return new FlexStagingDelegate(appEngineDirectory);
  }
```

### PR DESCRIPTION
**Pull request by chanseokoh**
Merging the feature branch into `master`. The branch is composed of just two previous PRs, without any addition or change:

- #2367: adding the JAR facet
- #2389: adding the context menu for Maven JAR projects and wiring up the JAR-packaging stager

As they become public, users will be able to add the JAR facet to any project. If it is added to, e.g., a Spring Boot Maven project, it can be deployed to flex.

(Additionally fixes #2391.)

### CODE REVIEW

**Comment by chanseokoh** [comment]
> Do we need a Maven listener for changes of <packaging>war</packaging> → <packaging>jar</packaging>? See .appengine.standard.java8's lifecycle-mapping.

IMO, I don't think we need or want to try anything here. Changing the packaging in `pom.xml` between `war` and `jar` only makes sense in a Spring Boot app, and even with the change in `pom.xml`, that's the only change. That is, the Spring Boot app itself and everything remains the same, always having `main()` intact, and its nature is still an executable JAR app. If it is not a Spring boot app, such a change will be a very rare event, mostly a one-time massive restructuring of the entire app to have `main()` in the end. Considering the rarity and the degree of destructiveness, I think it's better to let users drive the changes to ensure a sane state in the end, rather than we merely change the facet which would lead to a broken state anyway. That is, users should uninstall the WAR facet and WTP themselves, do proper restructuring of the codebase, and add the JAR facet back. It is inevitable a complex and long migration process, and I think this fully justifies the unnecessity of our attempt to something here.

**Comment by chanseokoh** [comment]
> re: 2 flex facets vs 1 flex facet with multiple versions, I've come to think that we should have a single facet with multiple versions. Instead of version, think of it as variant. I think this makes more sense as they're all variants of configuring the flexible environment, and indeed the app.yaml has env: flex. The version=1 doesn't look right, and using a single facet with multiple versions gets rid of the need for <conflict> entries.

I actually took another moment to think if I could do the way you mentioned (I didn't forget the Design Doc comment) when implementing this, and after some thought, I turned down the idea again — I didn't feel comfortable with it. By looking at all other facets, the versions were all referring to versions of specifications or technology levels: Java (1.6, 1.7, ...), EAR (5, 6, ...), Servlet, EJB, JSF, etc. So, although we could designate WAR and JAR as facet versions, I felt it does not fit well conceptually and a bit of abusing. Rather, we can think of the current flex env having a flex specification 1.0, and maybe later our platform could evolve to have flex specifications 2.0, 3.0, ... (`env: flex2` or `env: flex3`), although I admit I don't think that'll ever happen. In any case, I thought the version fits more nicely with these kinds of notions.

However, we can discuss more about this later

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui.test/build.properties [hunk 1]
```diff
@@ -4,3 +4,4 @@ bin.includes = META-INF/,\
                .
 javacSource=1.7
 javacTarget=1.7
+additional.bundles=org.apache.log4j
```
[the remark this answers is not in the document]
Comment (chanseokoh) [reply plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui.test/build.properties]: Can't remember this. I'm sure it was required in #2389. However, I tried removing this, and tests work fine. Removed.

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/plugin.xml [hunk 2]
```diff
@@ -80,22 +80,40 @@
         <and>
           <count value="1" />
           <iterate>
-            <reference definitionId="com.google.cloud.tools.eclipse.appengine.onlyInFlexProjects" />
+            <or>
+              <reference definitionId="com.google.cloud.tools.eclipse.appengine.onlyInFlexWarProjects" />
+              <reference definitionId="com.google.cloud.tools.eclipse.appengine.onlyInFlexJarMavenProjects" />
```
[the remark this answers is not in the document]
Comment (chanseokoh) [reply plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/plugin.xml]: Done.

File: plugins/com.google.cloud.tools.eclipse.appengine.facets/plugin.xml [hunk 3]
```diff
@@ -180,19 +181,71 @@
...
+        type="INSTALL">
       <delegate class="com.google.cloud.tools.eclipse.appengine.facets.FlexFacetInstallDelegate"/>
     </action>
     <action
-      id="com.google.cloud.tools.eclipse.appengine.facets.flex.uninstall.action"
-      facet="com.google.cloud.tools.eclipse.appengine.facets.flex"
-      type="UNINSTALL">
+        id="com.google.cloud.tools.eclipse.appengine.facets.flex.uninstall.action"
+        facet="com.google.cloud.tools.eclipse.appengine.facets.flex"
+        type="UNINSTALL">
```
[the remark this answers is not in the document]
Comment (chanseokoh) [reply plugins/com.google.cloud.tools.eclipse.appengine.facets/plugin.xml]: True. The currently flex WAR facet had it, so I just copied it.

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/flexible/FlexDeployCommandHandler.java [hunk 4]
```diff
@@ -39,14 +46,26 @@ protected DeployPreferencesDialog newDeployPreferencesDialog(Shell shell, IProje
...
+  protected StagingDelegate getStagingDelegate(IProject project) throws CoreException {
     String appYamlPath = new FlexDeployPreferences(project).getAppYamlPath();
     IFile appYaml = project.getFile(appYamlPath);
     if (!appYaml.exists()) {
       throw new IllegalStateException("Invalid app.yaml path from project preferences.");
     }
     IPath appEngineDirectory = appYaml.getParent().getLocation();
-    return new FlexStagingDelegate(appEngineDirectory);
+
+    IFacetedProject facetedProject = ProjectFacetsManager.create(project);
```
[the remark this answers is not in the document]
Comment (chanseokoh) [reply plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/flexible/FlexDeployCommandHandler.java]: It must be a faceted project at this point, which is after pressing _Deploy_. Added a pre-condition check.

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/flexible/FlexDeployCommandHandler.java [hunk 5]
```diff
@@ -39,14 +47,26 @@ protected DeployPreferencesDialog newDeployPreferencesDialog(Shell shell, IProje
...
+  protected StagingDelegate getStagingDelegate(IProject project) throws CoreException {
+    IFacetedProject facetedProject = ProjectFacetsManager.create(project);
+    Preconditions.checkNotNull(facetedProject);
+
     String appYamlPath = new FlexDeployPreferences(project).getAppYamlPath();
     IFile appYaml = project.getFile(appYamlPath);
-    if (!appYaml.exists()) {
-      throw new IllegalStateException("Invalid app.yaml path from project preferences.");
-    }
+    Preconditions.checkState(appYaml.exists(), "Invalid app.yaml path from project preferences.");
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/flexible/FlexDeployCommandHandler.java]: This should be a CoreException, not a state check. State checks are only for internal object state, not external file system state.

Comment (chanseokoh) [reply plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/flexible/FlexDeployCommandHandler.java]: Got it. Will update it.

Comment (briandealwis) [reply plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/flexible/FlexDeployCommandHandler.java]: Guava also offers `Verify`.

**Pull request merged**